docs/sphinxext/changelog.py

```diff
@@ -0,0 +1,152 @@
+from typing import Any, Dict, Iterable, List, Optional
+
+import requests
+from docutils import nodes
+from docutils.parsers.rst import Directive, directives
+
+
+class ChangelogDirective(Directive):
+
+    # defines the parameter the directive expects
+    # directives.unchanged means you get the raw value from RST
+    required_arguments = 0
+    optional_arguments = 0
+    final_argument_whitespace = True
+    option_spec = {"github": directives.unchanged, "pypi": directives.unchanged}
+    has_content = False
+    add_index = False
+
+    def run(self) -> Iterable[nodes.Node]:
+        # sett = self.state.document.settings
+        # env = self.state.document.settings.env
+
+        # gives you access to the parameter stored
+        # in the main configuration file (conf.py)
+        # config = env.config
+
+        # # we create a title and we add it to section
+        # section += nodes.title("bla")
+
+        config = self.state.document.settings.env.config
+        token = config.changelog_github_token
+        if not token:
+            # If we raise self.error here, that will keep people
+            # from building the doc without a token. We don't want that.
+            par = nodes.paragraph()
+            par += nodes.Text("(Changelog not built because ")
+            par += nodes.literal("", "changelog_github_token")
+            par += nodes.Text(" parameter is missing in conf.py)")
+            return [par]
+
+        owner_repo = self.extract_github_repo_name(self.options["github"])
+        releases = self.extract_releases(owner_repo, token)
+
+        pypi_name: Optional[str]
+        try:
+            pypi_name = self.extract_pypi_package_name(self.options["pypi"])
+        except KeyError:
+            pypi_name = None
+
+        result_nodes: List[nodes.Node] = []
+        for release in releases:
+            result_nodes.extend(
+                list(self.nodes_for_release(release, pypi_name=pypi_name))
+            )
+
+        return result_nodes
+
+    def extract_github_repo_name(self, url: str) -> str:
+        stripped_url = url.rstrip("/")
+        prefix, postfix = "https://github.com/", "/releases"
+        url_is_correct = stripped_url.startswith(prefix) and stripped_url.endswith(
+            postfix
+        )
+        if not url_is_correct:
+            raise self.error(
+                "Changelog needs a Github releases URL "
+                f"(https://github.com/:owner/:repo/releases). Received {url}"
+            )
+
+        return stripped_url[len(prefix) : -len(postfix)]  # noqa
```

Comment:
why the `noqa` ?

Reply:
https://github.com/psf/black#slices but you're right, it's not the correct fix for that.